src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitServiceParam_SimpleReturn_Snapshot.generated.txt

```diff
@@ -0,0 +1,303 @@
+            [global::System.Runtime.CompilerServices.CallerLineNumber]int lineNumber = 0)
+        {
+            return global::Microsoft.AspNetCore.Http.Generated.GeneratedRouteBuilderExtensionsCore.MapCore(
+                    endpoints,
+                    pattern,
+                    handler,
+                    PostVerb,
+                    filePath,
+                    lineNumber);
+        }
+        internal static global::Microsoft.AspNetCore.Builder.RouteHandlerBuilder MapPost(
+            this global::Microsoft.AspNetCore.Routing.IEndpointRouteBuilder endpoints,
+            [global::System.Diagnostics.CodeAnalysis.StringSyntax("Route")] string pattern,
+            global::System.Func<Microsoft.AspNetCore.Http.Generators.Tests.TestService?, System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Http.Generators.Tests.TestService>, string> handler,
+            [global::System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
+            [global::System.Runtime.CompilerServices.CallerLineNumber]int lineNumber = 0)
+        {
+            return global::Microsoft.AspNetCore.Http.Generated.GeneratedRouteBuilderExtensionsCore.MapCore(
+                    endpoints,
+                    pattern,
+                    handler,
+                    PostVerb,
+                    filePath,
+                    lineNumber);
+        }
+
+    }
+}
+
+namespace Microsoft.AspNetCore.Http.Generated
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Collections.ObjectModel;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Reflection;
+    using System.Threading.Tasks;
+    using System.IO;
+    using Microsoft.AspNetCore.Routing;
+    using Microsoft.AspNetCore.Routing.Patterns;
+    using Microsoft.AspNetCore.Builder;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Http.Metadata;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Extensions.FileProviders;
+    using Microsoft.Extensions.Primitives;
+
+    using MetadataPopulator = System.Func<System.Reflection.MethodInfo, Microsoft.AspNetCore.Http.RequestDelegateFactoryOptions?, Microsoft.AspNetCore.Http.RequestDelegateMetadataResult>;
+    using RequestDelegateFactoryFunc = System.Func<System.Delegate, Microsoft.AspNetCore.Http.RequestDelegateFactoryOptions, Microsoft.AspNetCore.Http.RequestDelegateMetadataResult?, Microsoft.AspNetCore.Http.RequestDelegateResult>;
+
+    file static class GeneratedRouteBuilderExtensionsCore
+    {
+
+        private static readonly Dictionary<(string, int), (MetadataPopulator, RequestDelegateFactoryFunc)> map = new()
+        {
+            [(@"TestMapActions.cs", 15)] = (
+               (methodInfo, options) =>
+                {
+                    Debug.Assert(options?.EndpointBuilder != null, "EndpointBuilder not found.");
```

Comment:
This Debug.Assert is kind of weird to generate in the user's project. What are they going to do if this assert ever fires?

Reply:
Actually, looking more, I think the issue is in the `MetadataPopulator` delegate:

```
System.Func<System.Reflection.MethodInfo, Microsoft.AspNetCore.Http.RequestDelegateFactoryOptions?, Microsoft.AspNetCore.Http.RequestDelegateMetadataResult>;
```

Why is this `RequestDelegateFactoryOptions?`? According to this assert, it isn't valid for `null` to be passed.

Reply:
If I recall correctly, a lot of the APIs in RDF take a nullable `RequestDelegateFactoryOptions` because we supported a scenario where the `RequestDelegateFactory` can be invoked outside of the route handler codepath, where we aren't guaranteed to have instantiated the options object. For the generator scenario, we know we are emitting for a route handler (e.g. `Delegate`) type so we can make the assertion here. The codepath being invoked is [here](https://github.com/dotnet/aspnetcore/blob/63a0251075468508022d900eb8e1e277c84bf4f5/src/Http/Routing/src/RouteEndpointDataSource.cs#L202-L208).

@halter73 Did I get the reasoning for the nullability on this type correct?

Reply:
Correct. But since `RouteHandlerServices.Map` is new API, we should consider changing `populateMetadata` from `Func<MethodInfo, RequestDelegateFactoryOptions?, RequestDelegateMetadataResult>` to `Action<EndpointBuilder>`. We could put the code to get it back into the right shape for `RouteEndpointBuilder.AddRouteHandler` and do our asserts inside of `RouteHandlerServices.Map`. It feels worth it since the generated code is pretty visible and the asserts are pretty ugly.

Reply:
Fair enough. I'll give this a try. From looking through the codepath a bit more, I think we might have to do the assert in `AddRouteHandler` and update a couple of the internal types. I can see about bringing this to API review tomorrow as well.